src/libcollections/str.rs

```diff
@@ -1634,30 +1634,53 @@ impl str {
     #[stable(feature = "unicode_case_mapping", since = "1.2.0")]
     pub fn to_lowercase(&self) -> String {
         let mut s = String::with_capacity(self.len());
-        for (i, c) in self[..].char_indices() {
-            if c == 'Σ' {
-                // Σ maps to σ, except at the end of a word where it maps to ς.
-                // This is the only conditional (contextual) but language-independent mapping
-                // in `SpecialCasing.txt`,
-                // so hard-code it rather than have a generic "condition" mechanism.
-                // See https://github.com/rust-lang/rust/issues/26035
-                map_uppercase_sigma(self, i, &mut s)
-            } else {
-                s.extend(c.to_lowercase());
+        let mut left = None;
+
+        // Try to collect slices of lower case characters to push into the
+        // result or extend with the lower case version if an upper case
+        // character is found.
+        for (i, ch) in self.char_indices() {
+            if !ch.is_lowercase() && ch.is_alphabetic() {
```

Comment:
I think adding `PartialEq<char>` to `ToLowercase` and `ToUppercase` would already allow the optimization to work.

From what I tested the performance improvement comes from using `push_str` instead of `extend`, because the `push_str` gets optimized to a `memcpy`, so the bigger the already properly cased substring is the faster it gets.

I can look into adding the `PartialEq<char>` impl and using it here if you want.